### PR TITLE
Add VSM to ValidationBehavior

### LINKED ISSUE #807: [Bug] Apply an explicit style to an Entry using TextValidationBehavior with InvalidStyle
### Description
from https://github.com/xamarin/XamarinCommunityToolkit/discussions/748

It seems impossible to apply an explicit style on an `Entry` which is using the `InvalidStyle` of `TextValidationBehavior`, you may say just append/add the explicit style to the `InvalidStyle` definition, but the issue is that the latter is only triggered upon text changed event I believe, thus the style won't be applied all the time.

Any ideas?

If I remove the Behaviors, `SomeEntryStyle` is applied, if I put back, only `InvalidEntryStyle` style will be applied when conditions met. the issue here is that `SomeEntryStyle`  should be applied regardless.

I am not sure if I am missing something or something needs to change/improve in xct that why I have put it as a bug for now.

### Steps to Reproduce
```
      <Entry Text="text"
                   Placeholder="palceholder"
                   Style="{StaticResource SomeEntryStyle}">
                <Entry.Behaviors>
                    <xct:EmailValidationBehavior DecorationFlags="Trim"
                                                 InvalidStyle="{StaticResource InvalidEntryStyle}"/>
                </Entry.Behaviors>
            </Entry>
```
```
    <ContentPage.Resources>
        <Style x:Key="SomeEntryStyle"
               TargetType="Entry">
            <Setter Property="TextColor" Value="White"/>
            <Setter Property="PlaceholderColor" Value="Blue"/>
            <Setter Property="BackgroundColor" Value="Purple"/>
            <Setter Property="Visual" Value="Material"/>
        </Style>

        <Style x:Key="InvalidEntryStyle"
               TargetType="Entry">
            <Setter Property="TextColor" Value="Red"/>
        </Style>
    </ContentPage.Resources>
```

### Basic Information

- Version with issue: 1.0.2
- Last known good version: -
- IDE: vs2019
- Platform Target Frameworks: All

### LINKED ISSUE #952: [Enhancement] ValidationBehavior VisualStates
## Summary
Support validation `VisualStates` with `ValidationBehavior`.

## API Changes
| VisualState Name | VisualStateGroup Name | Description                      |
|------------------|-----------------------|----------------------------------|
| Valid         | ValidationStates       | The value is valid.     |
| Invalid       | ValidationStates       | The value is not valid. |

## Intended Use Case
Define different visual appearances by `VSM`.

## Who Will Do The Work?
- [ ] I am willing to take this on myself
- [x] Just putting this out there for someone to pick up

### PR DESCRIPTION
**Pull request by Cfun1**
### Description of Change ###

Added support of visual state manger to ValidationBehavior.

### Bugs Fixed ###
- Fixes #952
- Fixes #807

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard

- [ ] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 1]
```diff
@@ -11,6 +12,9 @@ namespace Xamarin.CommunityToolkit.Behaviors.Internals
 	/// </summary>
 	public abstract class ValidationBehavior : BaseBehavior<VisualElement>
 	{
+		public const string ValidVisualState = "Valid";
+		public const string InvalidVisualState = "Invalid";
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:16]: Should we remove Public and make it readonly ?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 2]
```diff
@@ -291,12 +295,28 @@ async ValueTask UpdateStateAsync(bool isForced, CancellationToken? parentToken =
...
+			{
 				return;
+			}
+
+			if ((ValidStyle ?? InvalidStyle) != null)
+			{
+				View.Style = IsValid
+					? ValidStyle
+					: InvalidStyle;
+				return;     // either use VSM or styles?
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:308]: Any thoughts ?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 3]
```diff
@@ -291,12 +295,28 @@ async ValueTask UpdateStateAsync(bool isForced, CancellationToken? parentToken =
...
 
-			View.Style = IsValid
-				? ValidStyle
-				: InvalidStyle;
+			var vsmGroupsList = VisualStateManager.GetVisualStateGroups(View);
+			if (vsmGroupsList.Count > 0)
+			{
+				var vsmGroupsstates = vsmGroupsList[0].States;
+				if (vsmGroupsstates.All(x => x.Name.Equals(ValidVisualState) || x.Name.Equals(UnValidVisualState)))
+				{
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:316]: Do we need these checks? is it ok to use Linq? Do wee need to set the State to Null/default in some cases?

**Comment by pictos** [comment]
@Cfun1 can you check your code? Looks like that sometimes is missing... Could be something related to the cherry-pick

File: samples/XCT.Sample/Pages/Behaviors/CharactersValidationBehaviorPage.xaml [hunk 4]
```diff
@@ -1,69 +1,91 @@
...
-                                                                                  Source={x:Reference MaximumCharacterCountEntry}}"
-                                                  MinimumCharacterCount="{Binding Text,
-                                                                                  Source={x:Reference MinimumCharacterCountEntry}}" />
+                                                  ValidStyle="{StaticResource ValidEntryStyle}"
+                                                  MaximumCharacterCount="{Binding Text, Source={x:Reference MaximumCharacterCountEntry}}"
+                                                  MinimumCharacterCount="{Binding Text, Source={x:Reference MinimumCharacterCountEntry}}"/>
+            </Entry.Behaviors>
+        </Entry>
+
+        <Label Text="The same but stying handeled by visual state manager."/>
```
Comment (AndreiMisiukevich) [inline samples/XCT.Sample/Pages/Behaviors/CharactersValidationBehaviorPage.xaml:65]:
```suggestion
        <Label Text="The same but styling handled by visual state manager."/>
```

Comment (Cfun1) [reply samples/XCT.Sample/Pages/Behaviors/CharactersValidationBehaviorPage.xaml:65]: Thanks, corrected it but didn't push, and I blame a next cherry-pick :)

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 5]
```diff
@@ -291,12 +296,28 @@ async ValueTask UpdateStateAsync(bool isForced, CancellationToken? parentToken =
...
+			var vsmGroupsList = VisualStateManager.GetVisualStateGroups(View);
+			if (vsmGroupsList.Count > 0)
+			{
+				var vsmGroupsstates = vsmGroupsList[0].States;
+				if (vsmGroupsstates.All(x => x.Name.Equals(validVisualState) || x.Name.Equals(invalidVisualState)))
+				{
+					VisualStateManager.GoToState(View, IsValid ? validVisualState : invalidVisualState);
+				}
+			}
 		}
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:321]:
```csharp
		void UpdateStyle()
		{
			if (View == null)
				return;

			VisualStateManager.GoToState(View, IsValid ? validVisualState : invalidVisualState);

			if ((ValidStyle ?? InvalidStyle) == null)
				return;

			View.Style = IsValid
				? ValidStyle
				: InvalidStyle;
		}
```

Comment (Cfun1) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:321]: Checking is not required ?

Comment (AndreiMisiukevich) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:321]: I am not an expert in VisualStateManager)
So, I guess it's not required.

CC @pictos @jfversluis

Comment (jfversluis) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:321]: Sorry, checking what exactly?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 6]
```diff
@@ -11,6 +13,9 @@ namespace Xamarin.CommunityToolkit.Behaviors.Internals
 	/// </summary>
 	public abstract class ValidationBehavior : BaseBehavior<VisualElement>
 	{
+		const string validVisualState = "Valid";
+		const string invalidVisualState = "Invalid";
```
Comment (jfversluis) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:17]: Please make them public, people can refer to these to not have to rely on magic strings

**Review by jfversluis — CHANGES_REQUESTED** [review]

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs (hunk 5 above)
Comment (Cfun1) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:321]: @jfversluis Checking if dev has defined a VisualStateGroup with at least a VisualState having as name "Valid" or Invalid, or it is not require to check before invoking `VisualStateManager.GoToState(View, IsValid ? validVisualState : invalidVisualState);`

Comment (pictos) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:321]: If Style accepts `null` I believe that we don't the check. But the best case is to do a test and see how it goes.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 7]
```diff
@@ -11,6 +13,9 @@ namespace Xamarin.CommunityToolkit.Behaviors.Internals
 	/// </summary>
 	public abstract class ValidationBehavior : BaseBehavior<VisualElement>
 	{
+		public const string validVisualState = "Valid";
+		public const string invalidVisualState = "Invalid";
+
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:18]: Sorry, could you please make them capitalized, then Approved :)

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 8]
```diff
@@ -291,12 +295,28 @@ async ValueTask UpdateStateAsync(bool isForced, CancellationToken? parentToken =
 
 		void UpdateStyle()
 		{
-			if (View == null || (ValidStyle ?? InvalidStyle) == null)
+			if (View == null)
+			{
 				return;
+			}
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:301]:
```suggestion
			if (View == null)
				return;
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 9]
```diff
@@ -291,12 +295,28 @@ async ValueTask UpdateStateAsync(bool isForced, CancellationToken? parentToken =
...
+
+			var vsmGroupsList = VisualStateManager.GetVisualStateGroups(View);
+			if (vsmGroupsList.Count > 0)
+			{
+				var vsmGroupsstates = vsmGroupsList[0].States;
+				if (vsmGroupsstates.All(x => x.Name.Equals(ValidVisualState) || x.Name.Equals(InvalidVisualState)))
+				{
+					VisualStateManager.GoToState(View, IsValid ? ValidVisualState : InvalidVisualState);
+				}
+			}
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:319]:
```suggestion
			VisualStateManager.GoToState(View, IsValid ? validVisualState : invalidVisualState);

			if ((ValidStyle ?? InvalidStyle) == null)
				return;

			View.Style = IsValid ? ValidStyle : InvalidStyle;
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 10]
```diff
@@ -3,6 +3,7 @@
 using System.Threading.Tasks;
 using System.Windows.Input;
 using Xamarin.Forms;
+using System.Linq;
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:6]:
```suggestion
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 11]
```diff
@@ -291,12 +294,17 @@ async ValueTask UpdateStateAsync(bool isForced, CancellationToken? parentToken =
 
 		void UpdateStyle()
 		{
-			if (View == null || (ValidStyle ?? InvalidStyle) == null)
+			if (View == null)
+			{
+				return;
+			}
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:300]:
```suggestion
			if (View == null)
				return;
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 12]
```diff
@@ -291,12 +294,15 @@ async ValueTask UpdateStateAsync(bool isForced, CancellationToken? parentToken =
 
 		void UpdateStyle()
 		{
-			if (View == null || (ValidStyle ?? InvalidStyle) == null)
+			if (View == null)
+				return;
+
+			VisualStateManager.GoToState(View, IsValid ? validVisualState : invalidVisualState);
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:300]:
```suggestion
			if (View == null)
				return;

			VisualStateManager.GoToState(View, IsValid ? ValidVisualState : InvalidVisualState);
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs [hunk 13]
```diff
@@ -11,6 +11,9 @@ namespace Xamarin.CommunityToolkit.Behaviors.Internals
 	/// </summary>
 	public abstract class ValidationBehavior : BaseBehavior<VisualElement>
 	{
+		public const string ValidVisualState = "Valid";
+		public const string InvalidVisualState = "Invalid";
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:15]:
```suggestion

		public const string InvalidVisualState = "Invalid";
```

**Comment by AndreiMisiukevich** [comment]
@jfversluis your review is required cuz you requested changes)

**Review by AndreiMisiukevich — APPROVED** [review]

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs (hunk 5 above)
Comment (pictos) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators/ValidationBehavior.shared.cs:307]: @Cfun1 just to double-check, the current implementation works as expected with null values?

**Review by jfversluis — APPROVED** [review]

**Pull request merged**